text/0000-prelude-2021.md

```diff
@@ -0,0 +1,172 @@
+- Feature Name: `prelude_2021`
+- Start Date: 2021-02-16
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+A new prelude for the 2021 edition, featuring several extra traits.
+
+# Motivation
+[motivation]: #motivation
+
+While types and free functions can be added to the prelude independent of edition boundaries, the same is not true for traits. Adding a trait to the prelude can cause compatibility issues because calls to methods named the same as methods of the newly in scope traits can become ambiguous. Because editions are opt-in, they represent an opportunity to add new traits to the prelude.
+
+The expected outcome is to make the newly added traits more accessible to users, potentially stimulating their use compared to alternatives.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The compiler currently brings all items from `std::prelude::v1` into scope for each module (for code using `std`). Crates that declare usage of the 2021 edition will instead get items from `std::prelude::edition2021` imported into scope. With the implementation of this RFC, we will add `edition2015` and `edition2018` modules in `std::prelude`, and a crate configuring a particular edition will get the prelude appropriate for its edition. The `v1` module will stick around for compatibility reasons, although it might be deprecated in the future.
```

Comment:
Module names are snake_case by convention. This should be `std::prelude::edition_2021`.

Citation: `std::collections::hash_map`.

You might say, “but `v1`!”, but I consider that the correct spelling even in snake_case. `v_1` would look nasty.

In the very distant past, well before 1.0, there was some momentum for joinedlowerwords crate and module names, as Python does with its packages and modules, but that was steadily dismantled in favour of snake_case.

Reply:
I don't think that's correct, you're inferring from something that matches `[a-z]+_[a-z]+` that the valid handling for numeric tagging must be `[a-z]+(_([a-z]+|[0-9]+))*` when that does not exclude `[a-z]+[0-9]*(_[a-z]+[0-9]*)*` which is present in the compiler, extensively, as type and module names. `f32`, `f64`, `wasm32`, `aarch64`, and `x86_64` (a particularly amusing one, on reflection), are all in the compiler as part of module paths, and the compiler already has code that flags based on the name `edition2018` already. I won't say `edition_2021` is a wrong way to write it according to rustc conventions, just that module names tend to have special pressure on them to be curt and `edition2021` seems to be permissible usage.

Reply:
Is there anywhere else we use "edition_2021", "edition 2021", "edition-2021", or "edition2021"? If we do, we should aim to be consistent with that. In `Cargo.toml`, it's `edition = "2021"` which doesn't help guide one way or another.